### ISSUE TITLE
WaitForCommitAsync in write operations

### ISSUE THREAD

**Issue by potrusil-osi** [sakno/dotNext#25]
I'm trying to implement my own "reliable queue" on top of your implementation of the Raft protocol. But I don't fully understand how `PersistentState.WaitForCommitAsync` should be used when implementing write operations. My main confusion is the `timeout` parameter. If I set it to something like 2 seconds (which is the value in the RaftNode example), what is supposed to happen if the change cannot be committed during those 2 seconds, but it is committed right after `WaitForCommitAsync` returns false? (Actually, the same can happen with the `token` parameter.) I'd think that the inability to commit within a reasonable time should cancel the whole transaction, but in reality there is no connection between the operation side of things and the heartbeat (which commits entries). That means the write operation can fail, but it is eventually committed. How should I reason about this?

My second question is about how `PersistentState.WaitForCommitAsync` is used in the RaftNode project. There it waits on an index that is always one bigger than the last committed index:
``` csharp
            var commitIndex = GetLastIndex(true);
            await AppendAsync(new Int64LogEntry { Content = value, Term = Term }, token);
            await WaitForCommitAsync(commitIndex + 1L, timeout, token);
```
Is this only for the sake of simplicity? My understanding is that it would be better to wait on the index that is returned by `AppendAsync` the line above - that way the write operation, `UpdateValueAsync`, can be executed multiple time in parallel and this code would still work. Am I correct?
``` csharp
            var entryIndex = await AppendAsync(new Int64LogEntry { Content = value, Term = Term }, token);
            await WaitForCommitAsync(entryIndex, timeout, token);
```

**Comment by potrusil-osi** [comment]
After thinking this through a bit more I guess the described behavior of `WaitForCommitAsync` on timeout is OK and the solution is to implement the duplicates elimination. Am I thinking about it right?

**Comment by sakno** [comment]
`WaitForCommitAsync` doesn't provide transactional behavior. It's low-level API. Transactional behavior is out of scope of this Raft implementation. It's responsibility of the developer. The method allows you to ensure that the log entry at the specified index is marked as _committed_ by the majority of nodes. It means that the cluster accepts the log record in a consistent way and it cannot be removed from the log. Before that, the log entry has _uncommitted_ state which means that the only local node has this record in the log but all other nodes not. _Committing_ is a process of replication (distribution) of _uncommitted_ log entries by the leader across followers.

If `WaitForCommitAsync` is timed out then there is no guarantees to the caller about _committed_ state of the log entry. The entry can be rewritten if leader has changed accidentally. You have two ways in this case:
* Optimistic approach - attempt to add log entry again with the updated term and wait for commit. Of course you need to analyze the semantics of the log entry and ensure that this action will not overwrite already committed data
* Pessimistic approach - inform the caller that the write operation is timed out and ask to retry again

**Comment by sakno** [comment]
The selected `AppendAsync` overload is for simplicity, yes. it's not accurate but acceptable because commit index is increasing monotonically. This method is mostly used by Raft infrastructure for snapshot installation. For user code it's better to use more comprehensive overload:
```csharp
ValueTask<long> AppendAsync<TEntryImpl>(ILogEntryProducer<TEntryImpl> entries, CancellationToken token = default(CancellationToken))
```
This method returns exactly what you want.
